File: filterBillsAndFetchVotes.js
const fs = require('fs').promises;
const path = require('path');
const axios = require('axios');

async function processBillsData() {
  try {
    console.log('Starting to process assembly bills data...');
    
    // Get all files in the current directory
    const files = await fs.readdir(__dirname);
    
    // Filter files that start with 'assembly_bills_age' and end with '.json'
    const billsFiles = files.filter(file => 
      file.startsWith('assembly_bills_age') && file.endsWith('.json')
    );
    
    console.log(`Found ${billsFiles.length} assembly bills files to process.`);
    
    const filteredResults = [];
    
    for (const file of billsFiles) {
      console.log(`Processing file: ${file}`);
      
      try {
        const filePath = path.join(__dirname, file);
        const content = await fs.readFile(filePath, 'utf8');
        const jsonData = JSON.parse(content);
        
        if (!jsonData.data || !Array.isArray(jsonData.data)) {
          console.log(`Skipping ${file} - no data array found`);
          continue;
        }
        
        // Filter data where PROC_RESULT is '원안가결' or '수정가결'
        const filtered = jsonData.data.filter(item => 
          (item.PROC_RESULT === '원안가결' || item.PROC_RESULT === '수정가결')
          && Number(item.AGE) >= 17
        );
        
        // Extract BILL_ID and AGE
        const extracted = filtered.map(item => ({
          BILL_ID: item.BILL_ID,
          AGE: item.AGE
        }));
        
        filteredResults.push(...extracted);
        
        console.log(`Found ${filtered.length} matching records in ${file}`);
        
      } catch (error) {
        console.error(`Error processing file ${file}:`, error.message);
      }
    }
    
    // Remove duplicates based on BILL_ID
    const uniqueResults = filteredResults.filter((item, index, array) => 
      array.findIndex(i => i.BILL_ID === item.BILL_ID) === index
    );
    
    console.log(`Total filtered records: ${filteredResults.length}`);
    console.log(`Unique records after deduplication: ${uniqueResults.length}`);
    
    // Save results to JSON file
    const outputData = {
      total_count: uniqueResults.length,
      filtered_date: new Date().toISOString(),
      filter_criteria: "PROC_RESULT = '원안가결' OR '수정가결'",
      data: uniqueResults
    };
    
    const outputPath = path.join(__dirname, 'assembly_filtered_bills_passed.json');
    await fs.writeFile(outputPath, JSON.stringify(outputData, null, 2), 'utf8');
    
    console.log(`Results saved to: ${outputPath}`);
    console.log('Processing completed successfully!');
    
    return uniqueResults;
    
  } catch (error) {
    console.error('Error in processBillsData:', error);
    throw error;
  }
}

async function loadFilteredBillsAndCallAPI() {
  try {
    console.log('Loading filtered bills data...');
    
    // Load filtered_assembly_bills.json or filtered_bills_passed.json
    let filteredData;
    try {
      const filteredPath = path.join(__dirname, 'assembly_filtered_bills_passed.json');
      const content = await fs.readFile(filteredPath, 'utf8');
      filteredData = JSON.parse(content);
    } catch (error) {
      // Fallback to filtered_bills_passed.json if filtered_assembly_bills.json doesn't exist
      const filteredPath = path.join(__dirname, 'assembly_filtered_passed_bills.json');
      const content = await fs.readFile(filteredPath, 'utf8');
      filteredData = JSON.parse(content);
    }
    
    if (!filteredData.data || !Array.isArray(filteredData.data)) {
      throw new Error('No data array found in filtered bills file');
    }
    
    console.log(`Found ${filteredData.data.length} bills to process`);
    
    // Check for existing API results to avoid duplicates
    let existingResults = [];
    try {
      const existingPath = path.join(__dirname, 'assembly_bills_api_results.json');
      const existingContent = await fs.readFile(existingPath, 'utf8');
      const existingData = JSON.parse(existingContent);
      existingResults = existingData.results || [];
      console.log(`Found ${existingResults.length} existing API results`);
    } catch (error) {
      console.log('No existing API results found, starting fresh');
    }
    
    // Filter out bills that already have API results
    const billsToProcess = filteredData.data.filter(bill => {
      const exists = existingResults.some(result => 
        result.BILL_ID === bill.BILL_ID && result.AGE === bill.AGE
      );
      return !exists;
    });
    
    console.log(`After removing duplicates: ${billsToProcess.length} bills need API calls`);
    
    if (billsToProcess.length === 0) {
      console.log('All bills already have API results. No new calls needed.');
      return {
        summary: {
          total_bills_processed: existingResults.length,
          successful_calls: existingResults.filter(r => r.status === 'success').length,
          failed_calls: existingResults.filter(r => r.status === 'error').length,
          no_data_calls: existingResults.filter(r => r.status === 'no_data').length,
          processed_date: new Date().toISOString(),
          note: 'No new API calls made - all bills already processed'
        },
        results: existingResults
      };
    }
    
    const apiResults = [...existingResults]; // Start with existing results
    const API_KEY = '84d2bb829a0d413b97da6e9d2809db9e';
    const BASE_URL = 'https://open.assembly.go.kr/portal/openapi/nojepdqqaweusdfbi';
    
    // Process only new bills in batches
    const batchSize = 10;
    for (let i = 0; i < billsToProcess.length; i += batchSize) {
      const batch = billsToProcess.slice(i, i + batchSize);
      
      console.log(`Processing batch ${Math.floor(i / batchSize) + 1}/${Math.ceil(billsToProcess.length / batchSize)} (${batch.length} bills)`);
      
      const batchPromises = batch.map(async (bill) => {
        try {
          const url = `${BASE_URL}?KEY=${API_KEY}&Type=json&pIndex=1&pSize=1000&BILL_ID=${bill.BILL_ID}&AGE=${bill.AGE}`;
          
          console.log(`Calling API for BILL_ID: ${bill.BILL_ID}, AGE: ${bill.AGE}`);
          
          const response = await axios.get(url, {
            timeout: 30000, // 30 second timeout
          });
          
          if (response.data && response.data.nojepdqqaweusdfbi) {
            console.log(`✓ Success for BILL_ID: ${bill.BILL_ID}`);
            
            return {
              BILL_ID: bill.BILL_ID,
              AGE: bill.AGE,
              api_response: response.data.nojepdqqaweusdfbi,
              status: 'success',
              timestamp: new Date().toISOString()
            };
          } else {
            console.log(`- No data for BILL_ID: ${bill.BILL_ID}`);
            return {
              BILL_ID: bill.BILL_ID,
              AGE: bill.AGE,
              api_response: null,
              status: 'no_data',
              timestamp: new Date().toISOString()
            };
          }
        } catch (error) {
          console.error(`✗ Error for BILL_ID ${bill.BILL_ID}:`, error.message);
          return {
            BILL_ID: bill.BILL_ID,
            AGE: bill.AGE,
            api_response: null,
            status: 'error',
            error: error.message,
            timestamp: new Date().toISOString()
          };
        }
      });
      
      const batchResults = await Promise.all(batchPromises);
      apiResults.push(...batchResults);
      
      // Save intermediate results after each batch
      const intermediateData = {
        summary: {
          total_bills_processed: apiResults.length,
          successful_calls: apiResults.filter(result => result.status === 'success').length,
          failed_calls: apiResults.filter(result => result.status === 'error').length,
          no_data_calls: apiResults.filter(result => result.status === 'no_data').length,
          processed_date: new Date().toISOString(),
          batch_completed: Math.floor(i / batchSize) + 1
        },
        results: apiResults
      };
      
      try {
        const intermediatePath = path.join(__dirname, 'assembly_bills_api_results_temp.json');
        await fs.writeFile(intermediatePath, JSON.stringify(intermediateData, null, 2), 'utf8');
        console.log(`Intermediate results saved (${apiResults.length} total results)`);
      } catch (intermediateError) {
        console.warn(`Warning: Failed to save intermediate results:`, intermediateError.message);
        console.log(`Continuing with batch processing...`);
      }
      
      // Add delay between batches
      if (i + batchSize < billsToProcess.length) {
        console.log('Waiting 2 seconds before next batch...');
        await new Promise(resolve => setTimeout(resolve, 2000));
      }
    }
    
    // Final verification - check for any duplicates in results
    const uniqueResults = [];
    const seenBills = new Set();
    
    for (const result of apiResults) {
      const key = `${result.BILL_ID}_${result.AGE}`;
      if (!seenBills.has(key)) {
        seenBills.add(key);
        uniqueResults.push(result);
      } else {
        console.warn(`Duplicate found and removed: BILL_ID ${result.BILL_ID}, AGE ${result.AGE}`);
      }
    }
    
    // Compile results
    const compiledData = {
      summary: {
        total_bills_processed: uniqueResults.length,
        successful_calls: uniqueResults.filter(result => result.status === 'success').length,
        failed_calls: uniqueResults.filter(result => result.status === 'error').length,
        no_data_calls: uniqueResults.filter(result => result.status === 'no_data').length,
        processed_date: new Date().toISOString(),
        duplicates_removed: apiResults.length - uniqueResults.length
      },
      results: uniqueResults
    };
    
    // Save compiled results
    console.log('Saving compiled results...');
    const outputPath = path.join(__dirname, 'assembly_bills_api_results.json');
    
    try {
      // Convert to JSON string first to check if there are any issues
      let jsonString;
      try {
        jsonString = JSON.stringify(compiledData, null, 2);
        console.log(`JSON string length: ${jsonString.length} characters`);
      } catch (stringifyError) {
        console.warn('Warning: JSON stringify failed, trying without formatting:', stringifyError.message);
        try {
          jsonString = JSON.stringify(compiledData);
          console.log(`JSON string created without formatting. Length: ${jsonString.length} characters`);
        } catch (secondStringifyError) {
          console.error('Critical: JSON stringify failed completely:', secondStringifyError.message);
          console.log('Saving summary only and continuing...');
          
          // Save just the summary as fallback
          const summaryPath = path.join(__dirname, 'assembly_bills_summary_fallback.json');
          await fs.writeFile(summaryPath, JSON.stringify(compiledData.summary, null, 2), 'utf8');
          console.log(`Summary saved to: ${summaryPath}`);
          
          // Return the data even if saving failed
          console.log(`\nAPI calls completed despite saving issues!`);
          console.log(`Total processed: ${uniqueResults.length}`);
          console.log(`Successful: ${compiledData.summary.successful_calls}`);
          console.log(`Failed: ${compiledData.summary.failed_calls}`);
          console.log(`No data: ${compiledData.summary.no_data_calls}`);
          
          return compiledData;
        }
      }
      
      await fs.writeFile(outputPath, jsonString, 'utf8');
      console.log(`Results saved successfully to: ${outputPath}`);
      
      // Verify file was created and check its size
      try {
        const stats = await fs.stat(outputPath);
        console.log(`File size: ${stats.size} bytes`);
        
        // Try to read back the file to verify it's valid
        const readBack = await fs.readFile(outputPath, 'utf8');
        const parsed = JSON.parse(readBack);
        console.log(`Verification: File contains ${parsed.results.length} results`);
        
      } catch (verifyError) {
        console.warn('Warning: File verification failed but file was saved:', verifyError.message);
      }
      
    } catch (writeError) {
      console.warn('Warning: Main file save failed, trying fallback methods:', writeError.message);
      
      // Try saving a smaller test file first
      try {
        const testData = { test: 'data', timestamp: new Date().toISOString() };
        const testPath = path.join(__dirname, 'test_write.json');
        await fs.writeFile(testPath, JSON.stringify(testData, null, 2), 'utf8');
        console.log('Test file write successful');
        
        // Now try saving just the summary
        const summaryOnlyPath = path.join(__dirname, 'assembly_bills_summary.json');
        await fs.writeFile(summaryOnlyPath, JSON.stringify(compiledData.summary, null, 2), 'utf8');
        console.log(`Summary saved to: ${summaryOnlyPath}`);
        
        // Try saving results in chunks
        const chunkSize = 100;
        for (let i = 0; i < uniqueResults.length; i += chunkSize) {
          const chunk = uniqueResults.slice(i, i + chunkSize);
          const chunkPath = path.join(__dirname, `assembly_bills_chunk_${Math.floor(i / chunkSize) + 1}.json`);
          await fs.writeFile(chunkPath, JSON.stringify({ chunk_number: Math.floor(i / chunkSize) + 1, data: chunk }, null, 2), 'utf8');
          console.log(`Chunk ${Math.floor(i / chunkSize) + 1} saved to: ${chunkPath}`);
        }
        
      } catch (fallbackError) {
        console.warn('Warning: Even fallback saves failed, but processing continues:', fallbackError.message);
        // Don't throw - just continue with the process
      }
    }
    
    // Clean up temp file
    try {
      await fs.unlink(path.join(__dirname, 'assembly_bills_api_results_temp.json'));
      console.log('Temporary file cleaned up');
    } catch (cleanupError) {
      // Ignore cleanup errors
      console.log('Note: Temp file cleanup skipped (file may not exist)');
    }
    
    console.log(`\nAPI calls completed!`);
    console.log(`Total processed: ${uniqueResults.length}`);
    console.log(`Successful: ${compiledData.summary.successful_calls}`);
    console.log(`Failed: ${compiledData.summary.failed_calls}`);
    console.log(`No data: ${compiledData.summary.no_data_calls}`);
    if (compiledData.summary.duplicates_removed > 0) {
      console.log(`Duplicates removed: ${compiledData.summary.duplicates_removed}`);
    }
    
    return compiledData;
  } catch (error) {
    console.error('Error in loadFilteredBillsAndCallAPI:', error);
    throw error;
  }
}

// Run the script if this file is executed directly
if (require.main === module) {
  const args = process.argv.slice(2);
  
  if (args.includes('--filter-only')) {
    // Run original processing only
    processBillsData()
      .then(results => {
        console.log(`\nSummary: Processed ${results.length} unique bills that were passed.`);
        console.log(`To call APIs for these bills, run without --filter-only flag`);
      })
      .catch(error => {
        console.error('Script failed:', error);
        process.exit(1);
      });
  } else if (args.includes('--api-only')) {
    loadFilteredBillsAndCallAPI()
      .then(apiResults => {
        console.log(`\nAll processing completed successfully!`);
        console.log(`Final Summary: ${apiResults.summary.total_bills_processed} bills processed, ${apiResults.summary.successful_calls} successful API calls`);
      })
      .catch(error => {
        console.error('Processing failed:', error);
        process.exit(1);
      });
  } else {
    // Run both filtering and API calls by default
    console.log('Step 1: Filtering bills data...');
    processBillsData()
      .then(results => {
        console.log(`\nFiltering completed: Processed ${results.length} unique bills that were passed.`);
        console.log('\nStep 2: Starting API calls for filtered bills...\n');
        return loadFilteredBillsAndCallAPI();
      })
      .then(apiResults => {
        console.log(`\nAll processing completed successfully!`);
        console.log(`Final Summary: ${apiResults.summary.total_bills_processed} bills processed, ${apiResults.summary.successful_calls} successful API calls`);
      })
      .catch(error => {
        console.error('Processing failed:', error);
        process.exit(1);
      });
  }
}

module.exports = { processBillsData, loadFilteredBillsAndCallAPI };

File: filterBillsAndFetchVotes.py
import asyncio
import aiohttp
import json
import time
import os
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Any, Optional
import argparse
from dotenv import load_dotenv

load_dotenv()

class BillsFilterAndVoteFetcher:
    def __init__(self):
        self.base_dir = Path(__file__).parent
        self.api_key = os.getenv('API_KEY')
        if not self.api_key:
            raise ValueError('API_KEY is not set in environment variables. Please check your .env file.')
        self.base_url = 'https://open.assembly.go.kr/portal/openapi/nojepdqqaweusdfbi'

    def process_bills_data(self) -> List[Dict[str, Any]]:
        """Process assembly bills data and filter passed bills"""
        try:
            print('Starting to process assembly bills data...')
            
            # Get all files in the current directory
            files = [f for f in self.base_dir.iterdir() if f.is_file()]
            
            # Filter files that start with 'assembly_bills_age' and end with '.json'
            bills_files = [f for f in files if f.name.startswith('assembly_bills_age') and f.name.endswith('.json')]
            
            print(f'Found {len(bills_files)} assembly bills files to process.')
            
            filtered_results = []
            
            for file_path in bills_files:
                print(f'Processing file: {file_path.name}')
                
                try:
                    with open(file_path, 'r', encoding='utf-8') as f:
                        json_data = json.load(f)
                    
                    if 'data' not in json_data or not isinstance(json_data['data'], list):
                        print(f'Skipping {file_path.name} - no data array found')
                        continue
                    
                    # Filter data where PROC_RESULT is '원안가결' or '수정가결'
                    filtered = [
                        item for item in json_data['data']
                        if (item.get('PROC_RESULT') in ['원안가결', '수정가결'] and 
                            int(item.get('AGE', 0)) >= 17)
                    ]
                    
                    # Extract BILL_ID and AGE
                    extracted = [
                        {
                            'BILL_ID': item['BILL_ID'],
                            'AGE': item['AGE']
                        }
                        for item in filtered
                    ]
                    
                    filtered_results.extend(extracted)
                    
                    print(f'Found {len(filtered)} matching records in {file_path.name}')
                    
                except Exception as error:
                    print(f'Error processing file {file_path.name}: {error}')
            
            # Remove duplicates based on BILL_ID
            unique_results = []
            seen_bill_ids = set()
            
            for item in filtered_results:
                if item['BILL_ID'] not in seen_bill_ids:
                    seen_bill_ids.add(item['BILL_ID'])
                    unique_results.append(item)
            
            print(f'Total filtered records: {len(filtered_results)}')
            print(f'Unique records after deduplication: {len(unique_results)}')
            
            # Save results to JSON file
            output_data = {
                'total_count': len(unique_results),
                'filtered_date': datetime.now().isoformat(),
                'filter_criteria': "PROC_RESULT = '원안가결' OR '수정가결'",
                'data': unique_results
            }
            
            output_path = self.base_dir / 'assembly_filtered_bills_passed.json'
            with open(output_path, 'w', encoding='utf-8') as f:
                json.dump(output_data, f, ensure_ascii=False, indent=2)
            
            print(f'Results saved to: {output_path}')
            print('Processing completed successfully!')
            
            return unique_results
            
        except Exception as error:
            print(f'Error in process_bills_data: {error}')
            raise error

    def load_filtered_bills(self) -> Dict[str, Any]:
        """Load filtered bills data from JSON file"""
        try:
            # Try main filtered file first
            filtered_path = self.base_dir / 'assembly_filtered_bills_passed.json'
            if filtered_path.exists():
                with open(filtered_path, 'r', encoding='utf-8') as f:
                    return json.load(f)
            
            # Fallback to alternative filename
            filtered_path = self.base_dir / 'assembly_filtered_passed_bills.json'
            with open(filtered_path, 'r', encoding='utf-8') as f:
                return json.load(f)
                
        except Exception as error:
            raise Exception(f'Could not load filtered bills file: {error}')

    def load_existing_results(self) -> List[Dict[str, Any]]:
        """Load existing API results to avoid duplicates"""
        try:
            existing_path = self.base_dir / 'assembly_bills_api_results.json'
            with open(existing_path, 'r', encoding='utf-8') as f:
                existing_data = json.load(f)
                return existing_data.get('results', [])
        except:
            print('No existing API results found, starting fresh')
            return []

    async def call_vote_api(self, session: aiohttp.ClientSession, bill: Dict[str, Any]) -> Dict[str, Any]:
        """Call API for a single bill to get vote data"""
        try:
            url = f'{self.base_url}?KEY={self.api_key}&Type=json&pIndex=1&pSize=1000&BILL_ID={bill["BILL_ID"]}&AGE={bill["AGE"]}'
            
            print(f'Calling API for BILL_ID: {bill["BILL_ID"]}, AGE: {bill["AGE"]}')
            
            async with session.get(url, timeout=30) as response:
                response.raise_for_status()
                data = await response.json()
                
                if data and 'nojepdqqaweusdfbi' in data:
                    print(f'✓ Success for BILL_ID: {bill["BILL_ID"]}')
                    
                    return {
                        'BILL_ID': bill['BILL_ID'],
                        'AGE': bill['AGE'],
                        'api_response': data['nojepdqqaweusdfbi'],
                        'status': 'success',
                        'timestamp': datetime.now().isoformat()
                    }
                else:
                    print(f'- No data for BILL_ID: {bill["BILL_ID"]}')
                    return {
                        'BILL_ID': bill['BILL_ID'],
                        'AGE': bill['AGE'],
                        'api_response': None,
                        'status': 'no_data',
                        'timestamp': datetime.now().isoformat()
                    }
                    
        except Exception as error:
            print(f'✗ Error for BILL_ID {bill["BILL_ID"]}: {error}')
            return {
                'BILL_ID': bill['BILL_ID'],
                'AGE': bill['AGE'],
                'api_response': None,
                'status': 'error',
                'error': str(error),
                'timestamp': datetime.now().isoformat()
            }

    async def fetch_vote_data(self) -> Dict[str, Any]:
        """Load filtered bills and call API for vote data"""
        try:
            print('Loading filtered bills data...')
            
            filtered_data = self.load_filtered_bills()
            
            if 'data' not in filtered_data or not isinstance(filtered_data['data'], list):
                raise Exception('No data array found in filtered bills file')
            
            print(f'Found {len(filtered_data["data"])} bills to process')
            
            # Check for existing API results to avoid duplicates
            existing_results = self.load_existing_results()
            print(f'Found {len(existing_results)} existing API results')
            
            # Filter out bills that already have API results
            bills_to_process = []
            for bill in filtered_data['data']:
                exists = any(
                    result.get('BILL_ID') == bill['BILL_ID'] and 
                    result.get('AGE') == bill['AGE']
                    for result in existing_results
                )
                if not exists:
                    bills_to_process.append(bill)
            
            print(f'After removing duplicates: {len(bills_to_process)} bills need API calls')
            
            if not bills_to_process:
                print('All bills already have API results. No new calls needed.')
                return {
                    'summary': {
                        'total_bills_processed': len(existing_results),
                        'successful_calls': sum(1 for r in existing_results if r.get('status') == 'success'),
                        'failed_calls': sum(1 for r in existing_results if r.get('status') == 'error'),
                        'no_data_calls': sum(1 for r in existing_results if r.get('status') == 'no_data'),
                        'processed_date': datetime.now().isoformat(),
                        'note': 'No new API calls made - all bills already processed'
                    },
                    'results': existing_results
                }
            
            api_results = existing_results.copy()  # Start with existing results
            
            # Process bills in batches
            batch_size = 10
            connector = aiohttp.TCPConnector(limit=batch_size)
            timeout = aiohttp.ClientTimeout(total=30)
            
            async with aiohttp.ClientSession(connector=connector, timeout=timeout) as session:
                for i in range(0, len(bills_to_process), batch_size):
                    batch = bills_to_process[i:i + batch_size]
                    
                    print(f'Processing batch {i // batch_size + 1}/{(len(bills_to_process) + batch_size - 1) // batch_size} ({len(batch)} bills)')
                    
                    # Process batch concurrently
                    batch_tasks = [self.call_vote_api(session, bill) for bill in batch]
                    batch_results = await asyncio.gather(*batch_tasks, return_exceptions=True)
                    
                    # Handle results and exceptions
                    for result in batch_results:
                        if isinstance(result, Exception):
                            print(f'Batch processing error: {result}')
                        else:
                            api_results.append(result)
                    
                    # Save intermediate results after each batch
                    intermediate_data = {
                        'summary': {
                            'total_bills_processed': len(api_results),
                            'successful_calls': sum(1 for r in api_results if r.get('status') == 'success'),
                            'failed_calls': sum(1 for r in api_results if r.get('status') == 'error'),
                            'no_data_calls': sum(1 for r in api_results if r.get('status') == 'no_data'),
                            'processed_date': datetime.now().isoformat(),
                            'batch_completed': i // batch_size + 1
                        },
                        'results': api_results
                    }
                    
                    try:
                        intermediate_path = self.base_dir / 'assembly_bills_api_results_temp.json'
                        with open(intermediate_path, 'w', encoding='utf-8') as f:
                            json.dump(intermediate_data, f, ensure_ascii=False, indent=2)
                        print(f'Intermediate results saved ({len(api_results)} total results)')
                    except Exception as intermediate_error:
                        print(f'Warning: Failed to save intermediate results: {intermediate_error}')
                    
                    # Add delay between batches
                    if i + batch_size < len(bills_to_process):
                        print('Waiting 2 seconds before next batch...')
                        await asyncio.sleep(2)
            
            # Remove duplicates from final results
            unique_results = []
            seen_bills = set()
            
            for result in api_results:
                key = f"{result.get('BILL_ID')}_{result.get('AGE')}"
                if key not in seen_bills:
                    seen_bills.add(key)
                    unique_results.append(result)
                else:
                    print(f'Duplicate found and removed: BILL_ID {result.get("BILL_ID")}, AGE {result.get("AGE")}')
            
            # Compile final results
            compiled_data = {
                'summary': {
                    'total_bills_processed': len(unique_results),
                    'successful_calls': sum(1 for r in unique_results if r.get('status') == 'success'),
                    'failed_calls': sum(1 for r in unique_results if r.get('status') == 'error'),
                    'no_data_calls': sum(1 for r in unique_results if r.get('status') == 'no_data'),
                    'processed_date': datetime.now().isoformat(),
                    'duplicates_removed': len(api_results) - len(unique_results)
                },
                'results': unique_results
            }
            
            # Save compiled results
            self.save_results(compiled_data)
            
            print(f'\nAPI calls completed!')
            print(f'Total processed: {len(unique_results)}')
            print(f'Successful: {compiled_data["summary"]["successful_calls"]}')
            print(f'Failed: {compiled_data["summary"]["failed_calls"]}')
            print(f'No data: {compiled_data["summary"]["no_data_calls"]}')
            
            return compiled_data
            
        except Exception as error:
            print(f'Error in fetch_vote_data: {error}')
            raise error

    def save_results(self, compiled_data: Dict[str, Any]) -> None:
        """Save compiled results to file with error handling"""
        print('Saving compiled results...')
        output_path = self.base_dir / 'assembly_bills_api_results.json'
        
        try:
            # Try to save the main file
            with open(output_path, 'w', encoding='utf-8') as f:
                json.dump(compiled_data, f, ensure_ascii=False, indent=2)
            print(f'Results saved successfully to: {output_path}')
            
            # Verify file was created
            if output_path.exists():
                file_size = output_path.stat().st_size
                print(f'File size: {file_size} bytes')
                
                # Try to read back the file to verify it's valid
                with open(output_path, 'r', encoding='utf-8') as f:
                    parsed = json.load(f)
                print(f'Verification: File contains {len(parsed.get("results", []))} results')
            
        except Exception as write_error:
            print(f'Warning: Main file save failed: {write_error}')
            
            # Try saving just the summary as fallback
            try:
                summary_path = self.base_dir / 'assembly_bills_summary_fallback.json'
                with open(summary_path, 'w', encoding='utf-8') as f:
                    json.dump(compiled_data['summary'], f, ensure_ascii=False, indent=2)
                print(f'Summary saved to: {summary_path}')
                
                # Try saving results in chunks
                chunk_size = 100
                results = compiled_data.get('results', [])
                for i in range(0, len(results), chunk_size):
                    chunk = results[i:i + chunk_size]
                    chunk_path = self.base_dir / f'assembly_bills_chunk_{i // chunk_size + 1}.json'
                    chunk_data = {'chunk_number': i // chunk_size + 1, 'data': chunk}
                    with open(chunk_path, 'w', encoding='utf-8') as f:
                        json.dump(chunk_data, f, ensure_ascii=False, indent=2)
                    print(f'Chunk {i // chunk_size + 1} saved to: {chunk_path}')
                
            except Exception as fallback_error:
                print(f'Warning: Even fallback saves failed: {fallback_error}')
        
        # Clean up temp file
        try:
            temp_path = self.base_dir / 'assembly_bills_api_results_temp.json'
            if temp_path.exists():
                temp_path.unlink()
                print('Temporary file cleaned up')
        except:
            print('Note: Temp file cleanup skipped (file may not exist)')

def main():
    """Main function with command line argument parsing"""
    parser = argparse.ArgumentParser(description='Filter assembly bills and fetch vote data')
    parser.add_argument('--filter-only', action='store_true', 
                       help='Run filtering only, skip API calls')
    parser.add_argument('--api-only', action='store_true', 
                       help='Run API calls only, skip filtering')
    
    args = parser.parse_args()
    
    fetcher = BillsFilterAndVoteFetcher()
    
    if args.filter_only:
        # Run filtering only
        results = fetcher.process_bills_data()
        print(f'\nSummary: Processed {len(results)} unique bills that were passed.')
        print('To call APIs for these bills, run without --filter-only flag')
    
    elif args.api_only:
        # Run API calls only
        async def run_api_only():
            api_results = await fetcher.fetch_vote_data()
            print(f'\nAll processing completed successfully!')
            print(f'Final Summary: {api_results["summary"]["total_bills_processed"]} bills processed, {api_results["summary"]["successful_calls"]} successful API calls')
        
        asyncio.run(run_api_only())
    
    else:
        # Run both filtering and API calls by default
        async def run_both():
            print('Step 1: Filtering bills data...')
            results = fetcher.process_bills_data()
            print(f'\nFiltering completed: Processed {len(results)} unique bills that were passed.')
            print('\nStep 2: Starting API calls for filtered bills...\n')
            
            api_results = await fetcher.fetch_vote_data()
            print(f'\nAll processing completed successfully!')
            print(f'Final Summary: {api_results["summary"]["total_bills_processed"]} bills processed, {api_results["summary"]["successful_calls"]} successful API calls')
        
        asyncio.run(run_both())

if __name__ == "__main__":
    main()